### ISSUE TITLE
Change default nonlinearity to identity

### ISSUE THREAD

**Issue by benanne** [Lasagne/Lasagne#138]
In #133 three people spoke out in favour of changing the default nonlinearity for layers to the identity. I'm making this a separate issue so we can stay on topic there. What does everyone else think?

If we make this change we should probably raise a warning if no nonlinearity is specified, since it will break almost everyone's code.

As @f0k pointed out it's also valuable to add a layer that does nothing more than apply a nonlinearity, I think that's a good idea as well (regardless of whether we decide to change the default).

**Comment by craffel** [comment]
IMHO, Pro:
- No nonlinearity makes sense as a default because it's the simplest case.  Any other choice assumes that one nonlinearity is somehow more fundamental than the rest, which isn't something I'm comfortable saying.

Con:
- If the user doesn't know that the default is no nonlinearity, their network will suck.  So, raising a warning is probably smart.

**Comment by ebattenberg** [comment]
Let me just make the obvious joke by saying that setting the default *nonlinearity* to "linear" is kind of silly. ;)


> Con:
> 
> If the user doesn't know that the default is no nonlinearity, their network will suck. So, raising a warning is probably smart.

This is a good reason to stick with something nonlinear as the default.  Right now it seems that ReLU's are the de facto standard in deep neural networks, so they're a good choice for the default, and I appreciate not having to specify for every layer I create (except for, say, the output layer).  


> IMHO, Pro:
> 
> No nonlinearity makes sense as a default because it's the simplest case. Any other choice assumes that one nonlinearity is somehow more fundamental than the rest, which isn't something I'm comfortable saying.

This deals with the eventuality of a new de facto nonlinearity.  If we choose rectifiers as the default, and something else takes over in a year, then we'd want to change the default.  If we just say "identity" is the default, then that rationale will always be valid.  

Another reason to go to with a "linear" default is if the typical use case becomes one where people are using custom complex nonlinearities (e.g., parametrized nonlinearities) that are implemented in a separate layer.  

If I had to vote, I think I'd go for sticking with the ReLU as the default because that's the way it is now and it fits the most common use case (for now).

**Comment by craffel** [comment]
> Let me just make the obvious joke by saying that setting the default nonlinearity to "linear" is kind of silly. ;)

We could always call it `activation_function`!  (note - I am not actually advocating for this)

> This deals with the eventuality of a new de facto nonlinearity. If we choose rectifiers as the default, and something else takes over in a year, then we'd want to change the default. If we just say "identity" is the default, then that rationale will always be valid.

Yeah, that's exactly my point.  It's only relatively recently that ReLUs became the standard, and new purported better nonlinearities for deep networks are being proposed all the time.

**Comment by benanne** [comment]
Rectified linear units have been around for at least 4 years and have supplanted sigmoidal nonlinearities as the go-to nonlinearity for neural networks since then. Plenty of alternatives have been proposed, each with their own advantages and disadvantages, but so far none have been able to displace rectification. So I think this default is sensible and the probability of this changing within the next few years is fairly low.

Having rectification as the default is really convenient and it's an extremely common use case, so I'm in favour of pragmatism here. All it takes to remove it is passing `nonlinearity=None` anyway, which is easy enough to remember and short enough to type. Layers in a neural network are supposed to be nonlinear, so I think it doesn't make much sense to make them linear by default just so the code is a bit cleaner. 

I think the broader discussion is about whether we want to isolate each operation into a separate `Layer` subclass, or whether we want a `Layer` to represent what people think of as "a layer of a neural network". For me personally this definitely includes the nonlinearity. So it'd be great to maintain this mapping of "true layers" to `Layer` instances whenever it's feasible and practical.

**Comment by f0k** [comment]
> Let me just make the obvious joke by saying that setting the default nonlinearity to "linear" is kind of silly. ;)

That's right :) But we would probably default it to `None`, which makes sense.

> Layers in a neural network are supposed to be nonlinear [...]. I think the broader discussion is about whether we want [...] a Layer to represent what people think of as "a layer of a neural network". [...] For me personally this definitely includes the nonlinearity.

Same for me, I think it's good to have the nonlinearity included in the layer. Still I think having it default to `None` would lead to clearer code because it would force everybody to state the nonlinearity used. If over 90% of lasagne layers in the world and in the next 10 years have rectified linear units, then it may be more pragmatic to default it to the rectifier, but stating it explicitly is still cleaner.

What about a third solution of globally specifying the default nonlinearity somewhere? `lasagne.nonlinearities.default = ...`? This could default to `None` and be freely modifiable by users. This way people could rely on the shortcut if they wanted to, and we wouldn't have to predict trending nonlinearities.

**Comment by craffel** [comment]
> Still I think having it default to None would lead to clearer code because it would force everybody to state the nonlinearity used. If over 90% of lasagne layers in the world and in the next 10 years have rectified linear units, then it may be more pragmatic to default it to the rectifier, but stating it explicitly is still cleaner.

:+1:

> What about a third solution of globally specifying the default nonlinearity somewhere? lasagne.nonlinearities.default = ...? This could default to None and be freely modifiable by users. This way people could rely on the shortcut if they wanted to, and we wouldn't have to predict trending nonlinearities.

I was thinking about this.  It would add a little overhead, but might be the best way to make everyone happy.  In particular, because usually the same nonlinearity is used throughout a network, it would make the code for constructing a network quite a bit cleaner.

**Comment by benanne** [comment]
I think the `lasagne.nonlinearities.default` idea is not bad, but I would still prefer if it were rectification instead of none.

This is one case where I feel the cost of adhering to "explicit is better than implicit" is a little too high. This basically affects 90%+ of the use cases of the library and makes it more involved to create even a simple 2-layer neural net.

**Comment by f0k** [comment]
> This basically affects 90%+ of the use cases of the library and makes it more involved to create even a simple 2-layer neural net.

Well, it would change the code from:
```python
import lasagne

layer1 = ...
layer2 = ...
layer3 = ...
```
to
```python
import lasagne
lasagne.nonlinearities.default = lasagne.nonlinearities.rectify

layer1 = ...
layer2 = ...
layer3 = ...
```
I actually think this looks nice. The only trouble I have with a global variable is that it will lead to problems if you have multiple files setting it to different values in their preamble, so I'd like my lasagne best with `nonlinearity` simply defaulting to `None`. I'll still eat it with `nonlinearity=lasagne.nonlinearities.rectify`, though. (Sorry for the analogy.)

**Comment by benanne** [comment]
I think it's easy to forget... a neural net is pretty much worthless when it's linear, so I think it should be nonlinear by default. I really see no practical advantages to making layers linear by default.

I do agree that being able to control the default globally could save a lot of typing, so that's probably worth supporting. Same thing for a separate `NonlinearLayer` or something to that extent, that does nothing more than apply a nonlinearity.

However, I think it will be hard to come to a consensus on what should be the default, so maybe it's time for another vote?

**Comment by craffel** [comment]
> I actually think this looks nice. 

Agreed.  It adds a line of code and makes it much more explicit what the network is doing.

> The only trouble I have with a global variable is that it will lead to problems if you have multiple files setting it to different values in their preamble

I don't think this is so bad - I think users will/should in general do a `lasagne.nonlinearities.default` assignment immediately before constructing a network.  It's possible, I suppose, that in the course of constructing the network some function is called which changes `lasagne.nonlinearities.default`, but that's pretty unlikely and behavior I think is OK to allow.

> I think it's easy to forget...

We could also do something like `lasagne.nonlinearities.set_default(lasagne.nonlinearities.rectify)` and `lasagne.nonlinearities.get_default()` and keep track of two private variables in `lasagne.nonlinearities`: `__default` and `__default_set = False`.  If `get_default()` is called with `__default_set == False` then we could throw a warning - basically denoting that the user never set a default nonlinearity.  I think this is messy but it does make it a little more user-friendly.

> However, I think it will be hard to come to a consensus on what should be the default, so maybe it's time for another vote?

I think it partially depends on how it's implemented and what the examples look like; if all of the examples/docs include a line like `lasagne.nonlinearities.default = lasagne.nonlinearities.whatever` and/or we have a warning for when no default nonlinearity is set (as I proposed above), I think it's fine to have it default to `None`; otherwise I agree that it's dangerous to have it default to `None` because it would make it easy for users to create useless linear networks.

**Comment by ebenolson** [comment]
So far in my code, I've used two techniques (batch normalization and parametrized ReLU) that require the base nonlinearity to be `identity`. It would be nice to not have to explicitly declare that in each layer, either with a linear default or a global setting.

Personally I like the Torch style explicitness of declaring nonlinearities separately, and it's pretty compact if you compose them i.e.

    layer1 = ReLU(BatchNorm(ConvLayer(...)))

**Comment by benanne** [comment]
> Personally I like the Torch style explicitness of declaring nonlinearities separately, and it's pretty compact if you compose them

This is the discussion I mentioned earlier about whether we want to isolate each individual operation in its own layer:

> the broader discussion is about whether we want to isolate each operation into a separate `Layer` subclass, or whether we want a `Layer` to represent what people think of as "a layer of a neural network". I'm in favour of the latter. For me personally this definitely includes the nonlinearity. So it'd be great to maintain this mapping of "true layers" to `Layer` instances (or, if necessary, macro functions that create the appropriate stack of `Layer` classes) whenever it's feasible and practical.

I think we don't because then we're basically just reimplementing Theano on top of Theano, which seems pointless. I think a `Layer` instance should represent a layer of a neural network, whenever that is practical.

As I said before, I definitely support adding a way to change the default for all layers in one go, but I really cannot get behind changing it to be the identity by default.

> We could also do something like `lasagne.nonlinearities.set_default(lasagne.nonlinearities.rectify)` and `lasagne.nonlinearities.get_default()` and keep track of two private variables in `lasagne.nonlinearities`: `__default` and `__default_set = False`.  If `get_default()` is called with `__default_set == False` then we could throw a warning - basically denoting that the user never set a default nonlinearity.  I think this is messy but it does make it a little more user-friendly.

I feel like this is overcomplicating things, I don't like the idea of having to define getters and setters. We're setting a variable, so we should use the Python syntax for setting a variable, which is assignment.

**Comment by benanne** [comment]
As for batch normalization and parameterized relus, it's fairly trivial to write a simple 'macro' function that generates the appropriate stack of layers, for example:

```
def dense(l_in, *args, **kwargs):
    l_dense = lasagne.layers.DenseLayer(l_in, nonlinearity=None, *args, **kwargs)
    return ParameterizedReluLayer(l_dense)
```

 That way you also need to define it only once. Okay, maybe twice, once for convolutional layers and once for dense layers :) But that's also why I support a changeable default.

**Comment by f0k** [comment]
@bennane: Could you add the milestone here, please, so we don't forget?

**Comment by benanne** [comment]
done. I think I'm still against it though :p

**Comment by f0k** [comment]
> I think I'm still against it though :p

I've heard you use leaky relus now? Doesn't that prove the point that any default will become outdated after a while?

**Comment by benanne** [comment]
I used them once, for one problem, that doesn't mean I'm a convert :p It's more about the idea of having the most common use case be the default, and not having them be entirely linear by default at the very least to prevent what is likely to be a very common mistake.

**Comment by instagibbs** [comment]
I wonder if the user could be warned if he is constructing layers (like Dense/Conv/etc) and putting them together with only identity between them? Aside from more theoretical research it shouldn't be very common I don't think.

**Comment by benanne** [comment]
I'm still firmly in the 'rectify-as-default' camp, but if we do decide to change the default to linear, I don't think we should add checks like these instead as that will lead to some pretty messy code. The warning would probably trigger in several use cases where it's not appropriate. It's very hard to get something like that right.

**Comment by f0k** [comment]
Soo... I think we should keep the default nonlinearity as it is merely because too much time has passed. I still don't really like it, and I don't like the `if nonlinearity is None: nonlinearity = identity` check in each and every layer, but changing it now might break things for a lot of users (although I've taken care to always specify the nonlinearity explicitly in my code and in any Lasagne documentation).

Any objections against closing this issue?

**Comment by f0k** [comment]
On the mailing list, I've seen some instances of people instantiating DenseLayers without specifying a nonlinearity, but expecting them to be nonlinear. That's something we've suspected anyway, but which is difficult to automatically warn against without annoying people using linear layers for a reason. This is another reason to have a nonlinear default nonlinearity. Closing this now!
